### PR TITLE
vendor: implement --versioned-dirs

### PR DESCRIPTION
**Pull request by jsgf**
Implement `--explicit-version` from standalone cargo-vendor. This helps with vendoring
performance as it avoids redundantly deleting and re-copying already vendored packages.

For example, re-vendoring cargo's dependencies it makes a big difference in wallclock
time. For initial vendoring it makes no difference, but re-vendoring (ie, when most or all dependencies haven't changed) without explicit versions is actually slightly slower
(5.8s -> 6s), but with explicit versions it goes from 5.8s -> 1.6s.

Timings:

Without explicit versions, initial vendor
real	0m5.810s
user	0m0.924s
sys	0m2.491s

Re-vendor:
real	0m6.083s
user	0m0.937s
sys	0m2.654s

With explicit versions, initial vendor:
real	0m5.810s
user	0m0.937s
sys	0m2.461s

Re-vendor:
real	0m1.567s
user	0m0.578s
sys	0m0.967s

Its interesting to look at the syscall summary:

Without explicit versions:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 25.17    1.104699          18     59432      1065 openat
 19.86    0.871574          21     41156     13825 unlink
 13.64    0.598739           2    210510           lstat
  9.02    0.395948          29     13208           copy_file_range
  8.00    0.351242          11     30245           read
  6.36    0.279005           3     72487      4476 statx
  5.35    0.235027           6     37219           write
  4.02    0.176267           3     58368           close
```
with explicit versions:
```
 29.38    0.419068          15     27798     13825 unlink
 25.52    0.364021           1    209586           lstat
 20.67    0.294788          16     17967      1032 openat
 10.42    0.148586           4     35646           write
  3.53    0.050350           3     13825           chmod
  3.14    0.044786           2     16701      1622 statx
  2.19    0.031171           1     16936           close
  1.86    0.026538          24      1078           rmdir
```

Specifically, there are a lot fewer opens, copy_file_ranges, and unlinks.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! I'm a little surprised though that a motivator for this change is performance, vendoring is typically a pretty rare operation, right? This was originally added to `cargo vendor` specifically for the directory structure rather than performance

Since this is a proposal for a new stable feature, would it be possible to add tests for this as well, exercising this to make sure it works right?

**Comment by alexcrichton** [comment]
I think it'd also be good to write up some documentation to explain this feature and why it might be used.

**Comment by jsgf** [comment]
Yeah, performance is starting to be a blocker with 900+ package. It probably wouldn't be for normal use, but since I'm developing a tool for vendored package management using cargo vendor as an underpinning, I'm noticing it pretty strongly. I'm on PTO right now so I haven't checked to see how much this PR improves it, but I was surprised at how much difference it make to vendoring cargo's deps.

The main motivations are:
- Keep directory names stable. The current behaviour causes spurious deltas in source control for no-op changes (both for the vendored crates, and paths that reference them).
- Consistent naming. It's actually quite annoying that the most recent version can't be easily determined by eye when looking in the directory.
- Update performance, particularly for small changes. `cargo vendor` should be O(change) rather than O(total size of vendored set). I suspect most vendored package directories are suffixless (70% for cargo), so its currently closer to the latter than the former.

I'll add some tests and docs.

**Comment by jsgf** [comment]
Test and doc added. LMK if you think there's more to be tested/documented.

File: src/etc/man/cargo-vendor.1 [hunk 1]
```diff
@@ -54,13 +54,21 @@ vendored and synced to the output.
 \fB\-\-no\-delete\fP
 .RS 4
 Don\(cqt delete the "vendor" directory when vendoring, but rather keep all
-existing contents of the vendor directory
+existing contents of the vendor directory.
+.RE
+.sp
+\fB\-\-explicit\-version\fP
+.RS 4
+Normally versions are only added to disamiguate multiple versions of the same package.
```
Comment (alexcrichton) [inline src/etc/man/cargo-vendor.1]: "disambiguate"

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

File: src/etc/man/cargo-vendor.1 [hunk 2]
```diff
@@ -54,13 +54,21 @@ vendored and synced to the output.
...
+Normally versions are only added to disambiguate multiple versions of the same package.
+This option causes all directories in the "vendor" directory to be explicitly versioned,
+which makes it easier to track the history of vendored packages over time, and can help
+with the performance of re-vendoring when only a subset of the packages have changed.
 .RE
 .sp
 \fB\-\-respect\-source\-config\fP
 .RS 4
 Instead of ignoring \fB[source]\fP configuration by default in \fB.cargo/config\fP
-read it and use it when downloading crates from crates.io, for example
```
Comment (joshtriplett) [inline src/etc/man/cargo-vendor.1]: Please don't make unrelated changes in a PR like this.

File: src/bin/cargo/commands/vendor.rs [hunk 3]
```diff
@@ -27,6 +27,12 @@ pub fn cli() -> App {
                 .help("Respect `[source]` config in `.cargo/config`")
                 .multiple(true),
         )
+        .arg(
+            Arg::with_name("explicit-version")
+                .short("-x")
+                .long("explicit-version")
+                .help("Always include version in subdir name"),
```
Comment (joshtriplett) [inline src/bin/cargo/commands/vendor.rs]: Would you consider calling this something like `--version-dirs` or similar? "explicit" doesn't seem very self-explanatory when encountered outside the documentation.

Comment (jsgf) [reply src/bin/cargo/commands/vendor.rs]: This is the same option that `cargo-vendor` had. I'm fine with renaming it though.

File: src/etc/man/cargo-vendor.1 (hunk 2 above)
Comment (jsgf) [reply src/etc/man/cargo-vendor.1]: I split them into their own commit. I can make a separate PR if you want, but it seems a bit much for a 2 char doc change.

File: src/bin/cargo/commands/vendor.rs (hunk 3 above)
Comment (joshtriplett) [reply src/bin/cargo/commands/vendor.rs]: Ah, I didn't realize that.

If you think there's value in providing a compatible option, you could provide this as an alias. I do think we should have a self-documenting option name though.

File: src/etc/man/cargo-vendor.1 (hunk 2 above)
Comment (joshtriplett) [reply src/etc/man/cargo-vendor.1]: Having them in their own commit suffices, sure.

File: src/bin/cargo/commands/vendor.rs (hunk 3 above)
Comment (jsgf) [reply src/bin/cargo/commands/vendor.rs]: OK, I'm inclined to drop the short parameter since `-x` wouldn't make sense, and it doesn't seem like a common enough feature to consume another letter for.

**Comment by jsgf** [comment]
I changed the option to `--versioned-dirs`

File: src/etc/man/cargo-vendor.1 [hunk 4]
```diff
@@ -54,13 +54,21 @@ vendored and synced to the output.
...
-existing contents of the vendor directory
+existing contents of the vendor directory.
+.RE
+.sp
+\fB\-\-versioned\-dirs\fP
+.RS 4
+Normally versions are only added to disambiguate multiple versions of the same package.
+This option causes all directories in the "vendor" directory to be versioned,
+which makes it easier to track the history of vendored packages over time, and can help
+with the performance of re-vendoring when only a subset of the packages have changed.
```
Comment (ehuss) [inline src/etc/man/cargo-vendor.1]: This needs to go in the `.adoc` file, otherwise it will be lost when the documentation is regenerated.  There is some documentation at https://github.com/rust-lang/cargo/tree/master/src/doc on how the docs are generated.

**Comment by jsgf** [comment]
@joshtriplett @ehuss Documentation updated, and I think all comments addressed.
